smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -95,88 +121,50 @@ public void generateRequestSerializers(GenerationContext context) {
-                return shape.getMember().accept(this);
-            }
-
-            @Override
-            public Void mapShape(MapShape shape) {
-                generateFunctionSignature(shape, (c, s) -> serializeDocumentMap(c, s.asMapShape().get()));
-                return shape.getValue().accept(this);
-            }
-
-            @Override
-            public Void structureShape(StructureShape shape) {
-                generateFunctionSignature(shape, (c, s) -> serializeDocumentStructure(c, s.asStructureShape().get()));
-                shape.getAllMembers().values().forEach(member -> member.accept(this));
-                return null;
-            }
-
-            @Override
-            public Void unionShape(UnionShape shape) {
-                generateFunctionSignature(shape, (c, s) -> serializeDocumentUnion(c, s.asUnionShape().get()));
-                shape.getAllMembers().values().forEach(member -> member.accept(this));
-                return null;
-            }
-        }));
+        serializingShapes.forEach(shape -> shape.accept(new ShapeSerializingVisitor(context, true)));
     }
 
     @Override
     public void generateResponseDeserializers(GenerationContext context) {
         TopDownIndex topDownIndex = context.getModel().getKnowledge(TopDownIndex.class);
-        for (OperationShape operation : topDownIndex.getContainedOperations(context.getService())) {
+        OperationIndex operationIndex = context.getModel().getKnowledge(OperationIndex.class);
+        Walker shapeWalker = new Walker(context.getModel().getKnowledge(NeighborProviderIndex.class).getProvider());
+
+        // Track the shapes we need to generate sub-serializers for.
+        Set<Shape> deserializingShapes = new TreeSet<>();
+        Set<StructureShape> deserializingErrors = new TreeSet<>();
+        List<OperationShape> containedOperations = topDownIndex.getContainedOperations(context.getService()).stream()
+                .sorted()
+                .collect(Collectors.toList());
+        for (OperationShape operation : containedOperations) {
             OptionalUtils.ifPresentOrElse(
                     operation.getTrait(HttpTrait.class),
                     httpTrait -> generateOperationDeserializer(context, operation, httpTrait),
                     () -> LOGGER.warning(String.format(
                             "Unable to generate %s protocol response bindings for %s because it does not have an "
                             + "http binding trait", getName(), operation.getId())));
+
+            operationIndex.getOutput(operation)
+                    .ifPresent(output -> deserializingShapes.addAll(shapeWalker.walkShapes(output).stream()
+                            // Don't generate a sub-serializer for the actual output shape.
+                            .filter(s -> !output.equals(s))
+                            .collect(Collectors.toSet())));
```

Comment:
Rather than materializing this into a set, you could just forEach and do deserializingShapes::add